storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -2150,43 +2161,136 @@ xtrabackup_io_throttling(void)
+    hash_table_t* table,
+    xb_filter_entry_t** result
+)
+{
+    xb_filter_entry_t* found = NULL;
+    HASH_SEARCH(name_hash, table, ut_fold_string(name),
+            xb_filter_entry_t*,
+            found, (void) 0,
+            !strcmp(found->name, name));
+
+    if (found && result) {
+        *result = found;
+    }
+    return (found != NULL);
+}
+
 /************************************************************************
-Checks if a given table name matches any of specifications in the --tables or
---tables-file options.
+Checks if a given table name matches any of specifications given in
+regex_list or tables_hash.
 
-@return TRUE on match. */
+@return TRUE on match or both regex_list and tables_hash are empty.*/
 static my_bool
-check_if_table_matches_filters(const char *name)
+check_if_table_matches_filters(const char *name,
+    const regex_list_t& regex_list,
+    hash_table_t* tables_hash)
 {
-	int 			regres;
-	xb_filter_entry_t*	table;
-	xb_regex_list_node_t*	node;
+	if (regex_list.empty() && !tables_hash) {
+		return(FALSE);
+	}
 
-	if (UT_LIST_GET_LEN(regex_list)) {
-		/* Check against regular expressions list */
-		for (node = UT_LIST_GET_FIRST(regex_list); node;
-		     node = UT_LIST_GET_NEXT(regex_list, node)) {
-			regres = xb_regexec(&node->regex, name, 1,
-					    tables_regmatch, 0);
-			if (regres != REG_NOMATCH) {
+	if (regex_list_check_match(regex_list, name)) {
+		return(TRUE);
+	}
 
-				return(TRUE);
-			}
-		}
+	if (tables_hash && find_filter_in_hashtable(name, tables_hash, NULL)) {
+		return(TRUE);
 	}
 
-	if (tables_hash) {
-		HASH_SEARCH(name_hash, tables_hash, ut_fold_string(name),
-			    xb_filter_entry_t*,
-			    table, (void) 0,
-			    !strcmp(table->name, name));
-		if (table) {
+	return FALSE;
+}
+
+enum skip_database_check_result {
+    DATABASE_SKIP,
+    DATABASE_SKIP_SOME_TABLES,
+    DATABASE_DONT_SKIP,
+    DATABASE_DONT_SKIP_UNLESS_EXPLICITLY_EXCLUDED,
+};
+
+/************************************************************************
+Checks if a database specified by name should be skipped from backup based on
+the --databases, --databases_file or --databases_exclude options.
 
-			return(TRUE);
+@return TRUE if entire database should be skipped,
+        FALSE otherwise.
+*/
+static
+skip_database_check_result
+check_if_skip_database(
+    const char* name  /*!< in: path to the database */
+)
+{
+	/* There are some filters for databases, check them */
+	xb_filter_entry_t*	database = NULL;
+
+	if (databases_exclude_hash &&
+		find_filter_in_hashtable(name, databases_exclude_hash, &database) &&
+		!database->has_tables) {
+		/* Database is found and there are no tables specified, skip the db. */
+		return DATABASE_SKIP;
+	}
+
+	if (databases_include_hash) {
+		if (!find_filter_in_hashtable(name, databases_include_hash, &database)) {
+		/* Database isn't found, skip the database */
+			return DATABASE_SKIP;
+		} else if (database->has_tables) {
+			return DATABASE_SKIP_SOME_TABLES;
```

Comment:
`check_if_skip_database` will be invoked with `name="db"` and if hash table contains only `"db.t1"` record, then it will not be found, right? Or it is possible that in some cases `check_if_skip_database` invoked with table name?